File: JavaFactorial/src/de/luschny/math/factorial/FactorialBoitenSplit.java
// Copyright (C) 2004-2009 Peter Luschny, MIT License applies.
// See http://en.wikipedia.org/wiki/MIT_License
// Visit http://www.luschny.de/math/factorial/FastFactorialFunctions.htm
// Comments mail to: peter(at)luschny.de
package de.luschny.math.factorial;

import de.luschny.math.arithmetic.Xint;

public class FactorialBoitenSplit implements IFactorialFunction {

    public FactorialBoitenSplit() {
    }

    @Override
    public String getName() {
        return "BoitenSplit       ";
    }

    @Override
    public Xint factorial(int n) {
        if (n < 0) {
            throw new ArithmeticException("Factorial: n has to be >= 0, but was " + n);
        }

        if (n < 2) {
            return Xint.ONE;
        }

        Xint p = Xint.ONE;
        Xint r = Xint.ONE;

        // log2n = floor(log2(n));
        int log2n = 31 - Integer.numberOfLeadingZeros(n);
        int h = 0, shift = 0, k = 1;

        while (h != n) {
            shift += h;
            h = n >>> log2n--;
            int high = (h & 1) == 1 ? h : h - 1;

            while (k != high) {
                k += 2;
                p = p.multiply(k);
            }
            r = r.multiply(p);
        }
        return r.shiftLeft(shift);
    }
} // endOfFactorialSplitBoiten

File: SilverFactorial64/Sharith/Factorial/FactorialBoitenSplit.cs
// -------- ToujoursEnBeta
// Author & Copyright : Peter Luschny
// License: LGPL version 3.0 or (at your option)
// Creative Commons Attribution-ShareAlike 3.0
// Comments mail to: peter(at)luschny.de
// Created: 2010-03-01

namespace Sharith.Factorial 
{
    using XInt = Arithmetic.XInt;
    using XMath = MathUtils.XMath;

    public class BoitenSplit : IFactorialFunction 
    {
        public string Name => "BoitenSplit         ";

        public XInt Factorial(int n)
        {
            if (n < 0)
            {
                throw new System.ArgumentOutOfRangeException(
                    this.Name + ": " + nameof(n) + " >= 0 required, but was " + n);
            }

            if (n < 2) return XInt.One;

            var p = XInt.One;
            var r = XInt.One;

            int h = 0, shift = 0, k = 1;
            var log2N = XMath.FloorLog2(n);

            while (h != n)
            {
                shift += h;
                h = n >> log2N--;
                var high = (h & 1) == 1 ? h : h - 1;

                while (k != high)
                {
                    k += 2;
                    p *= k;
                }
                r *= p;
            }

            return r << shift;
        }
    }
} // endOfFactorialBoitenSplit
